src/sync/arc.rs

```diff
@@ -74,8 +74,14 @@ impl<T> Arc<T> {
 
     /// Consumes the `Arc`, returning the wrapped pointer.
     pub fn into_raw(this: Self) -> *const T {
-        let ptr = &*this as *const _;
+        let ptr = Self::as_ptr(&this);
         mem::forget(this);
+        ptr
+    }
+
+    /// Provides a raw pointer to the data.
+    pub fn as_ptr(this: &Self) -> *const T {
+        let ptr = &*this as *const _;
         ptr as *const T
```

Comment:
This should really go through the std `Arc` method for getting a pointer instead of just casting.

Reply:
What do you mean and why?
If you mean `std::sync::Arc::as_ptr(this)` - it is also just casting.

Reply:
What about
```
    /// Provides a raw pointer to the data.
    pub fn as_ptr(this: &Self) -> *const T {
        &this.inner.value
    }
```
?

Reply:
No, `std::sync::Arc::as_ptr` is not just casting. It's using `ptr::addr_of_mut!` to get a pointer to the `data` field, which may or may not be the first field. Even if was just casting, we can't do the same because it's an implementation detail that could change in the future.

As for `&this.inner.value`, that runs into problems with provenance. See the comment from the `as_ptr` in the standard library:
```text
pub fn as_ptr(this: &Self) -> *const T {
    let ptr: *mut ArcInner<T> = NonNull::as_ptr(this.ptr);

    // SAFETY: This cannot go through Deref::deref or RcBoxPtr::inner because
    // this is required to retain raw/mut provenance such that e.g. `get_mut` can
    // write through the pointer after the Rc is recovered through `from_raw`.
    unsafe { ptr::addr_of_mut!((*ptr).data) }
}
```
Together with #232, the following is ok:
```suggestion
    pub fn as_ptr(this: &Self) -> *const T {
        std::sync::Arc::as_ptr(&this.inner) as *const T
```